jax/_src/lax/control_flow/for_loop.py

```diff
@@ -273,3 +281,95 @@ def _swap_transpose(g, ref, x, *idx):
+# (`Ref`s should never be outputs from jaxprs). We'd like to convert that jaxpr
+# into a "pure" jaxpr that takes in and outputs values and no longer has the
+# `State` effect.
+
+def discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any]) -> Tuple[core.Jaxpr, List[Any]]:
+  """Converts a jaxpr that takes in `Ref`s into one that doesn't."""
+  in_avals = [core.ShapedArray(v.aval.shape, v.aval.dtype)
+              if type(v.aval) is ShapedArrayRef
+              else v.aval for v in jaxpr.invars]
+  eval_jaxpr = lu.wrap_init(partial(_eval_jaxpr_discharge_state, jaxpr, consts))
+  new_jaxpr, _ , new_consts = pe.trace_to_jaxpr_dynamic(eval_jaxpr, in_avals)
+  return new_jaxpr, new_consts
+
+def _dynamic_index(x, idx):
+  if not idx: return x
+  ndim = len(x.shape)
+  starts = [*idx] + [lax.full_like(idx[0], 0, shape=())] * (ndim - len(idx))
+  sizes = (1,) * len(idx) + x.shape[len(idx):]
+  out = lax.dynamic_slice(x, starts, sizes)
+  return out.reshape(x.shape[len(idx):])
+
+def _dynamic_update_index(x, idx, val):
+  if not idx: return val
+  ndim = len(x.shape)
+  starts = [*idx] + [lax.full_like(idx[0], 0, shape=())] * (ndim - len(idx))
+  update = val.reshape((1,) * len(idx) + x.shape[len(idx):])
+  return lax.dynamic_update_slice(x, update, starts)
+
+def _eval_jaxpr_discharge_state(jaxpr: core.Jaxpr, consts: Sequence[Any],
+                                *args: Any):
+  env: Dict[core.Var, Any] = {}
+
+  def read(v: core.Atom) -> Any:
+    if type(v) is core.Literal:
+      return v.val
+    assert isinstance(v, core.Var)
+    return env[v]
+
+  def write(v: core.Var, val: Any) -> None:
+    env[v] = val
+
+  map(write, jaxpr.constvars, consts)
+  # Here some args may correspond to `Ref` avals but they'll be treated like
+  # regular values in this interpreter.
+  map(write, jaxpr.invars, args)
+
+  for eqn in jaxpr.eqns:
+    in_vals = map(read, eqn.invars)
+    if eqn.primitive is get_p:
+       # `y <- x[i]` becomes `y = ds x i`
+      x, *idx = in_vals
+      write(eqn.outvars[0], _dynamic_index(x, idx))
+    elif eqn.primitive is swap_p:
+      # `z, x[i] <- x[i], val` becomes:
+      #    z = ds x i
+      #    x = dus x i val
+      x, val, *idx = in_vals
+      write(eqn.outvars[0], _dynamic_index(x, idx))
+      assert isinstance(eqn.invars[0], core.Var)
+      write(eqn.invars[0], _dynamic_update_index(x, idx, val))
+    elif eqn.primitive is addupdate_p:
+      # `x[i] += val` becomes:
+      #    y = ds x ii
+      #    z = y + val
+      #    x = dus x i z
+      x, val, *idx = in_vals
+      ans = _dynamic_update_index(x, idx, val + _dynamic_index(x, idx))
+      assert isinstance(eqn.invars[0], core.Var)
+      write(eqn.invars[0], ans)
+    else:
+      # Default primitive rule, similar to `core.eval_jaxpr`. Note that here
+      # we assume any higher-order primitives inside of the jaxpr are *not*
+      # stateful.
+      # TODO(sharadmv): enable discharging state in higher order primitives
```

Comment:
I think we've set things up so that no higher-order primitives can be stateful, since the only place the state effect is allowed is in the body of for_loop and that discharges it. We should think about releasing the state effect from its current prison someday, but in the meantime I think a code TODO is too narrow than that big design question, so maybe just remove it for now.

Reply:
Done!